core/state/transient_storage.go

```diff
@@ -0,0 +1,55 @@
+// Copyright 2022 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package state
```

Comment:
Just for curiosity. This implementation is much simpler than the EIP's demo implementation: https://eips.ethereum.org/EIPS/eip-1153#reference-implementation

the demo has `Journal` and `Checkpoints` to amortize the overhead. but here it uses `Copy()` as a snapshot method. Do you know why? @NazariiDenha Are these two implementations basically the same in runtime.

Reply:
I guess that geth uses a trick that combines the snapshot with the transaction's sub-context snapshot, but I have not confirmed this consumption.